File: examples/03-composition/gateway/main.py
import logging
import os
from contextlib import asynccontextmanager
from typing import Optional

import grpc
import httpx
import strawberry
from fastapi import FastAPI
from opentelemetry import trace
from opentelemetry.exporter.otlp.proto.http.trace_exporter import OTLPSpanExporter
from opentelemetry.instrumentation.fastapi import FastAPIInstrumentor
from opentelemetry.sdk.resources import Resource
from opentelemetry.sdk.trace import TracerProvider
from opentelemetry.sdk.trace.export import BatchSpanProcessor
from strawberry.dataloader import DataLoader
from strawberry.fastapi import GraphQLRouter

import inventory_pb2
import inventory_pb2_grpc

logging.basicConfig(level=logging.INFO, format="%(asctime)s %(levelname)s %(message)s")
log = logging.getLogger("gateway")

ORDER_API_URL = os.getenv("ORDER_API_URL", "http://order-api:8081")
INVENTORY_ADDR = os.getenv("INVENTORY_ADDR", "inventory:50051")
OTEL_ENDPOINT = os.getenv("OTEL_EXPORTER_OTLP_ENDPOINT", "http://lgtm:4318")

resource = Resource.create({"service.name": "gateway"})
provider = TracerProvider(resource=resource)
provider.add_span_processor(
    BatchSpanProcessor(OTLPSpanExporter(endpoint=f"{OTEL_ENDPOINT}/v1/traces"))
)
trace.set_tracer_provider(provider)
tracer = trace.get_tracer(__name__)

http_client: Optional[httpx.AsyncClient] = None
grpc_channel: Optional[grpc.aio.Channel] = None
inventory_stub: Optional[inventory_pb2_grpc.InventoryStub] = None


async def batch_load_stock(skus: list[str]) -> list[Optional[int]]:
    with tracer.start_as_current_span("batch_load_stock", attributes={"skus": str(skus)}):
        reply = await inventory_stub.GetStockBatch(
            inventory_pb2.GetStockBatchRequest(skus=skus)
        )
        stock_map = {item.sku: item.available for item in reply.items}
        log.info("DataLoader batched %d skus in one gRPC call", len(skus))
        return [stock_map.get(sku) for sku in skus]


@strawberry.type
class OrderType:
    id: str
    sku: str
    quantity: int
    status: str

    @strawberry.field
    async def stock(self, info: strawberry.types.Info) -> Optional[int]:
        return await info.context["stock_loader"].load(self.sku)


@strawberry.type
class Query:
    @strawberry.field
    async def orders(self, info: strawberry.types.Info, limit: int = 50) -> list[OrderType]:
        with tracer.start_as_current_span("fetch_orders_rest"):
            resp = await http_client.get(f"{ORDER_API_URL}/orders", params={"limit": limit})
            resp.raise_for_status()
            data = resp.json()
        return [OrderType(**o) for o in data]

    @strawberry.field
    async def order(self, info: strawberry.types.Info, id: str) -> Optional[OrderType]:
        with tracer.start_as_current_span("fetch_order_rest"):
            resp = await http_client.get(f"{ORDER_API_URL}/orders/{id}")
            if resp.status_code == 404:
                return None
            resp.raise_for_status()
            data = resp.json()
        return OrderType(**data)


schema = strawberry.Schema(query=Query)


async def get_context():
    return {"stock_loader": DataLoader(load_fn=batch_load_stock)}


graphql_router = GraphQLRouter(schema, context_getter=get_context)


@asynccontextmanager
async def lifespan(app: FastAPI):
    global http_client, grpc_channel, inventory_stub
    http_client = httpx.AsyncClient(timeout=10.0)
    grpc_channel = grpc.aio.insecure_channel(INVENTORY_ADDR)
    inventory_stub = inventory_pb2_grpc.InventoryStub(grpc_channel)
    log.info("gateway started")
    yield
    await http_client.aclose()
    await grpc_channel.close()


app = FastAPI(title="GraphQL Gateway", lifespan=lifespan)
FastAPIInstrumentor.instrument_app(app)
app.include_router(graphql_router, prefix="/graphql")


@app.get("/healthz")
async def healthz():
    return {"status": "ok"}

File: examples/03-composition/cpp/gateway/src/main.cpp
#include <drogon/drogon.h>
#include <grpcpp/grpcpp.h>
#include "inventory.grpc.pb.h"
#include <iostream>
#include <regex>
#include <set>

using namespace drogon;

static std::string g_order_api_url;
static std::unique_ptr<inventory::Inventory::Stub> g_inv_stub;

static std::unordered_map<std::string, int> batch_load_stock(const std::set<std::string>& skus) {
    inventory::GetStockBatchRequest req;
    for (auto& s : skus) req.add_skus(s);

    inventory::GetStockBatchReply reply;
    grpc::ClientContext ctx;
    auto status = g_inv_stub->GetStockBatch(&ctx, req, &reply);

    std::unordered_map<std::string, int> result;
    if (status.ok()) {
        for (const auto& item : reply.items()) {
            result[item.sku()] = item.available();
        }
        std::cerr << "DataLoader batched " << skus.size()
                  << " skus in one gRPC call" << std::endl;
    }
    return result;
}

static void handle_orders_query(const std::string& fields_str, int limit,
                                std::function<void(const HttpResponsePtr&)> cb) {
    bool want_stock = fields_str.find("stock") != std::string::npos;

    auto client = HttpClient::newHttpClient(g_order_api_url);
    auto req = HttpRequest::newHttpRequest();
    req->setMethod(Get);
    req->setPath("/orders?limit=" + std::to_string(limit));

    client->sendRequest(req,
        [fields_str, want_stock, cb = std::move(cb)](ReqResult result, const HttpResponsePtr& resp) {
            if (result != ReqResult::Ok) {
                Json::Value err;
                err["errors"][0]["message"] = "order-api unreachable";
                cb(HttpResponse::newHttpJsonResponse(err));
                return;
            }

            Json::CharReaderBuilder rb;
            Json::Value orders_arr;
            std::string body_str(resp->body());
            std::istringstream ss(body_str);
            Json::parseFromStream(rb, ss, &orders_arr, nullptr);

            std::unordered_map<std::string, int> stock_map;
            if (want_stock) {
                std::set<std::string> skus;
                for (const auto& o : orders_arr) skus.insert(o["sku"].asString());
                stock_map = batch_load_stock(skus);
            }

            Json::Value items(Json::arrayValue);
            for (const auto& o : orders_arr) {
                Json::Value item;
                if (fields_str.find("id") != std::string::npos) item["id"] = o["id"];
                if (fields_str.find("sku") != std::string::npos) item["sku"] = o["sku"];
                if (fields_str.find("quantity") != std::string::npos) item["quantity"] = o["quantity"];
                if (fields_str.find("status") != std::string::npos) item["status"] = o["status"];
                if (want_stock) {
                    auto it = stock_map.find(o["sku"].asString());
                    item["stock"] = (it != stock_map.end()) ? it->second : 0;
                }
                items.append(item);
            }

            Json::Value response;
            response["data"]["orders"] = items;
            cb(HttpResponse::newHttpJsonResponse(response));
        });
}

static void handle_order_query(const std::string& id, const std::string& fields_str,
                               std::function<void(const HttpResponsePtr&)> cb) {
    auto client = HttpClient::newHttpClient(g_order_api_url);
    auto req = HttpRequest::newHttpRequest();
    req->setMethod(Get);
    req->setPath("/orders/" + id);

    client->sendRequest(req,
        [fields_str, cb = std::move(cb)](ReqResult result, const HttpResponsePtr& resp) {
            Json::Value response;
            if (result != ReqResult::Ok || resp->statusCode() == k404NotFound) {
                response["data"]["order"] = Json::nullValue;
                cb(HttpResponse::newHttpJsonResponse(response));
                return;
            }

            Json::CharReaderBuilder rb;
            Json::Value o;
            std::string body_str(resp->body());
            std::istringstream ss(body_str);
            Json::parseFromStream(rb, ss, &o, nullptr);

            Json::Value item;
            if (fields_str.find("id") != std::string::npos) item["id"] = o["id"];
            if (fields_str.find("sku") != std::string::npos) item["sku"] = o["sku"];
            if (fields_str.find("quantity") != std::string::npos) item["quantity"] = o["quantity"];
            if (fields_str.find("status") != std::string::npos) item["status"] = o["status"];
            if (fields_str.find("stock") != std::string::npos) {
                std::set<std::string> skus = {o["sku"].asString()};
                auto stock_map = batch_load_stock(skus);
                auto it = stock_map.find(o["sku"].asString());
                item["stock"] = (it != stock_map.end()) ? it->second : 0;
            }
            response["data"]["order"] = item;
            cb(HttpResponse::newHttpJsonResponse(response));
        });
}

int main() {
    auto env_api = std::getenv("ORDER_API_URL");
    g_order_api_url = env_api ? env_api : "http://order-api:8081";

    auto env_inv = std::getenv("INVENTORY_ADDR");
    std::string inv_addr = env_inv ? env_inv : "inventory:50051";
    auto channel = grpc::CreateChannel(inv_addr, grpc::InsecureChannelCredentials());
    g_inv_stub = inventory::Inventory::NewStub(channel);

    std::cerr << "gateway started" << std::endl;

    app().setLogLevel(trantor::Logger::kInfo);
    app().addListener("0.0.0.0", 8080);

    app().registerHandler("/healthz",
        [](const HttpRequestPtr&, std::function<void(const HttpResponsePtr&)>&& cb) {
            Json::Value r;
            r["status"] = "ok";
            cb(HttpResponse::newHttpJsonResponse(r));
        },
        {Get});

    app().registerHandler("/graphql",
        [](const HttpRequestPtr& req, std::function<void(const HttpResponsePtr&)>&& cb) {
            auto body = req->getJsonObject();
            if (!body || !body->isMember("query")) {
                Json::Value err;
                err["errors"][0]["message"] = "missing query field";
                cb(HttpResponse::newHttpJsonResponse(err));
                return;
            }

            std::string query = (*body)["query"].asString();

            std::regex orders_re(R"re(orders\s*(?:\(\s*limit\s*:\s*(\d+)\s*\))?\s*\{([^}]*)\})re");
            std::regex order_re(R"re(order\s*\(\s*id\s*:\s*\\?"([^"\\]+)\\?"\s*\)\s*\{([^}]*)\})re");
            std::smatch match;

            if (std::regex_search(query, match, orders_re)) {
                int limit = match[1].matched ? std::stoi(match[1].str()) : 50;
                std::string fields = match[2].str();
                handle_orders_query(fields, limit, std::move(cb));
            } else if (std::regex_search(query, match, order_re)) {
                std::string id = match[1].str();
                std::string fields = match[2].str();
                handle_order_query(id, fields, std::move(cb));
            } else {
                Json::Value err;
                err["errors"][0]["message"] = "unsupported query";
                cb(HttpResponse::newHttpJsonResponse(err));
            }
        },
        {Post});

    app().run();
    return 0;
}
